### DISCUSSION TITLE
Can Detekt be used to create an annotation to enforce parameters in DSLs?

### DISCUSSION THREAD

**Discussion by elect86** [detekt/detekt#4886] [Q&A] [answered]
As titled, I'm looking for a non-android solution to [this](https://blog.kotlin-academy.com/kotlin-dsl-know-your-limits-2deaef1bab66), which uses an Android linter library:

```
@DSLint
class Person {

    @set:DSLMandatory
    var name: String
}
```

![image](https://miro.medium.com/max/1400/1*lzanVUqZD7kfnX3cuiBNOQ.png)

I hope this might be possible in compile time through the definition of a custom config definition or something similar, but I'm not familiar with static analyzers, so this is all speculation

**Comment by cortinico** [comment, marked as answer]
Hey @elect86 

> As titled, I'm looking for a non-android solution to [this](https://blog.kotlin-academy.com/kotlin-dsl-know-your-limits-2deaef1bab66), which uses an Android linter library:

That's totally doable. You can create your custom rule using this template https://github.com/detekt/detekt-custom-rule-template

And inspect the DSL calls + verify that certain parameters are declared or not. It can be totally custom (i.e. you hardcode the business logic inside the body of the rule) or you can provide some degree of configurability via external params or so.

**Comment by elect86** [comment]
Ok so, I [pushed](https://github.com/elect86/DslMandatory) (I modified one of the tests):
```
        annotation class DslMandatory(text: String)
        class Person(
            @DslMandatory
            var name: String,
            var age: Int)
        fun person(configure: Person.() -> Unit) = Person().configure()
        fun main() {
            person { 
            }
        }
```

My first idea is to work around `visitLambdaExpression` and then from there detect if there is a receiver and if that receiver has one the fields annotated by `DslMandatory`

But I have no idea how to do this.. any idea?

**Reply by cortinico** [reply]
For future reference: I pushed a draft version of the rule here https://github.com/elect86/DslMandatory/commit/189e8b6dde82d0e274c24816e844337267ce0bce
